src/python-common/ceph/fs/earmarking.py

```diff
@@ -0,0 +1,110 @@
+"""
+Module: CephFS Volume Earmarking
+
+This module provides the `CephFSVolumeEarmarking` class, which is designed to manage the earmarking
+of subvolumes and subvolume groups within a CephFS filesystem. The earmarking mechanism allows
+administrators to tag specific subvolumes with identifiers that indicate their intended use
+such as NFS or SMB, ensuring that only one file service is assigned to a particular subvolume
+or subvolume group at a time. This is crucial to prevent data corruption in environments where
+mixed protocol support (NFS and SMB) is not yet available.
+
+Key Features:
+- **Set Earmark**: Assigns an earmark to a subvolume.
+- **Get Earmark**: Retrieves the existing earmark of a subvolume, if any.
+- **Remove Earmark**: Removes the earmark from a subvolume, making it available for reallocation.
+- **Validate Earmark**: Ensures that the earmark follows the correct format and only uses
+supported top-level scopes.
+"""
+
+import errno
+import enum
+import logging
+from typing import Optional
+
+log = logging.getLogger(__name__)
+
+XATTR_SUBVOLUME_EARMARK_NAME = 'user.ceph.subvolume.earmark'
+
+
+class EarmarkScope(enum.Enum):
+    NFS = "nfs"
+    SMB = "smb"
+
+
+class EarmarkException(Exception):
+    def __init__(self, error_code: int, error_message: str) -> None:
+        self.errno = error_code
+        self.error_str = error_message
+
+    def to_tuple(self) -> tuple:
+        return self.errno, "", self.error_str
+
+    def __str__(self) -> str:
+        return f"{self.errno} ({self.error_str})"
+
+
+class CephFSVolumeEarmarking:
+    def __init__(self, fs, path: str) -> None:
+        self.fs = fs
+        self.path = path
+
+    def _handle_cephfs_error(self, e: Exception, action: str) -> None:
+        if isinstance(e, ValueError):
+            raise EarmarkException(errno.EINVAL, f"Invalid earmark specified: {e}")
+        elif isinstance(e, OSError):
+            log.error(f"Error {action} earmark: {e}")
+            raise EarmarkException(-e.errno, e.strerror)
+        else:
+            log.error(f"Unexpected error {action} earmark: {e}")
+            raise EarmarkException(errno.EIO, "Unexpected error")
+
+    def _validate_earmark(self, earmark: str) -> bool:
+        """
+        Validates that the earmark string is either empty or composed of parts separated by scopes,
+        with the top-level scope being either 'nfs' or 'smb'.
+
+        :param earmark: The earmark string to validate.
+        :return: True if valid, False otherwise.
+        """
+        if not earmark:
+            return True
+
+        if earmark in [EarmarkScope.NFS.value, EarmarkScope.SMB.value]:
```

Comment:
Try:
```
if earmark in EarmarkScope.__members__.keys():
```

Reply:
But this will give us key only right wheres we need values of those keys..so maybe simple like 
`earmark in (scope.value for scope in EarmarkTopScope)`

Reply:
Yeah, I messed that example up. Your suggestion is fine. The important point is not to assume that the members of the enum is fixed.